compiler/noirc_frontend/src/parser/parser/generics.rs

```diff
@@ -0,0 +1,295 @@
+use noirc_errors::Span;
+
+use crate::{
+    ast::{
+        GenericTypeArg, GenericTypeArgs, IntegerBitSize, Signedness, UnresolvedGeneric,
+        UnresolvedGenerics, UnresolvedType, UnresolvedTypeData,
+    },
+    parser::{labels::ParsingRuleLabel, ParserErrorReason},
+    token::{Keyword, Token, TokenKind},
+};
+
+use super::{parse_many::separated_by_comma, Parser};
+
+impl<'a> Parser<'a> {
+    /// Generics = ( '<' GenericsList? '>' )?
+    ///
+    /// GenericsList = Generic ( ',' Generic )* ','?
+    pub(super) fn parse_generics(&mut self) -> UnresolvedGenerics {
+        if !self.eat_less() {
+            return Vec::new();
+        }
+
+        self.parse_many(
+            "generic parameters",
+            separated_by_comma().until(Token::Greater),
+            Self::parse_generic_in_list,
+        )
+    }
+
+    fn parse_generic_in_list(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(generic) = self.parse_generic() {
+            Some(generic)
+        } else {
+            self.expected_label(ParsingRuleLabel::GenericParameter);
+            None
+        }
+    }
+
+    /// Generic
+    ///     = VariableGeneric
+    ///     | NumericGeneric
+    ///     | ResolvedGeneric
+    fn parse_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(generic) = self.parse_variable_generic() {
+            return Some(generic);
+        }
+
+        if let Some(generic) = self.parse_numeric_generic() {
+            return Some(generic);
+        }
+
+        if let Some(generic) = self.parse_resolved_generic() {
+            return Some(generic);
+        }
+
+        None
+    }
+
+    /// VariableGeneric = identifier
+    fn parse_variable_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(ident) = self.eat_ident() {
+            return Some(UnresolvedGeneric::Variable(ident));
+        }
+
+        None
+    }
+
+    /// NumericGeneric = 'let' identifier ':' Type
+    fn parse_numeric_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if !self.eat_keyword(Keyword::Let) {
+            return None;
+        }
```

Comment:
if we return an Option from eat_keyword then we could `?` here

Reply:
I agree that in many cases we return `None` when a keyword isn't there. I'll start changing these to return `Option`.

Reply:
I started making this change but the places where the return value is used as a bool are much more than those where we could early return with `?`, so I'm not sure it's worth it (it would make the code also read a bit worse, with `if self.eat_something().is_some()` instead of `if self.eat_something()`